cms/server/contest/templates/register.html

```diff
@@ -33,120 +33,212 @@
+            $("#confirmed-mask").show();
+            $("#show-username").text(data);
+        }).fail(function(data) {
+            // Check if this username exists
+            if (data.status === 404) {
+                $("#join #username-input").addClass("error");
+                $("#join #username-input span.help-block").text("{% trans %}No such user with this username found.{% endtrans %}");
+                $("#join #username").focus();
+            }
+            // Check if the password is correct
+            if (data.status === 403) {
+                $("#join #password-input").addClass("error");
+                $("#join #password-input span.help-block").text("{% trans %}The given password is not correct, try again.{% endtrans %}");
+                $("#join #password").focus();
+            }
+            // Check if the participation is not already added
+            if (data.status === 409) {
+                $("#join #username-input").addClass("error");
+                $("#join #username-input span.help-block").text("{% trans %}This user is already registered in the contest.{% endtrans %}");
+                $("#join #username").focus();
+            }
+        }).always(function() {
+            $("#join #submit-button-loading").hide();
+        });
+
+        e.preventDefault();
+    });
 });
 </script>
 
-<div class="login_container">
-    <div class="login_box hero-unit" style="margin-top: -300px;">
-
-        <div id="signup-mask">
-            <h1>{% trans %}New user{% endtrans %}</h1>
-            <p>{% trans %}Please fill in the fields to register{% endtrans %}</p>
-            <form id="signup" class="form-horizontal" action="{{ contest_url('register') }}">
-                {{ xsrf_form_html|safe }}
-                {% set next_page = handler.get_argument("next", none) %}
-                {% if next_page is not none %}
-                <input type="hidden" name="next" value="{{ next_page }}">
-                {% endif %}
-                <fieldset>
-                    <div class="control-group">
-                        <label class="control-label" for="first_name">{% trans %}First name{% endtrans %}</label>
-                        <div class="controls">
-                            <input required minlength="1" maxlength="{{ MAX_INPUT_LENGTH }}" type="text" class="input-xlarge" name="first_name" id="first_name">
+<div class="register_box hero-unit">
+    <div id="signup-mask">
+        <h1>{% trans %}Register{% endtrans %}</h1>
+        <p>{% trans %}Please fill in the fields to register{% endtrans %}</p>
+        <ul class="nav nav-tabs">
+            <li class="active"><a href="#tab_join" data-toggle="tab">Join Contest</a></li>
+            <li><a href="#tab_register" data-toggle="tab">New User</a></li>
+        </ul>
+        <div class="tab-content">
+            <div class="tab-pane active" id="tab_join">
+                <form id="join" class="form-horizontal" action="{{ contest_url('register') }}">
```

Comment:
Thanks!

I think that if there was one form instead of two, then it would be an improvement. How this is achieved not so essential. Newer Bootstrap does have styled radio buttons, but for Bootstrap 2 we have tabs and pills, which also look fine. (Perhaps pills look a little better for this use case.)

Reply:
Note also that, if you are going to implement a single-form solution, you'll likely find that the team field is in an unsatisfactory position (above login fields) if other fields are disabled/hidden. In this case, I think, it would be fine to just swap the two form halves that are separated by `<hr>`.

Reply:
Just applied the comments a few minutes ago.

Here's a list of changes:
1. A single html form is used
1. Tabs replaced with pills
1. Extra inputs hide in "join" use case
1. <hr> hides and 2 halves of form(say credentials and profile part) get swapped in "join" use case

Reply:
Thanks!

Sorry for not being more specific on (4). I meant that the form halves could be physically swapped in the HTML template source code, without Javascript. But, actually, the current implementation looks pretty good or even better, so it can be left as is 🙂